<a href="https://colab.research.google.com/github/Khuks/Predicting-Aggregate-Liquidity-For-A-Central-Bank-Using-Machine-Learning-Algorithms-/blob/main/Predicting_Aggregate_Liquidity_For_A_Central_Bank_Using_Machine_Learning_Algorithms_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Data Preprocessing

## Import Libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

## Load in the dataset

In [3]:
# load the dataset using pandas dataframe 
dataset=pd.read_excel('Data Preparation.xlsx')

#### Display the first five rows 

In [4]:
dataset.head()

Date  Total liquidity  Cash and Deposits with CBS  \
0   January 2011              NaN                         NaN   
1  February 2011              NaN                         NaN   
2     March 2011        1625736.0                    434426.0   
3     April 2011              NaN                         NaN   
4       May 2011              NaN                         NaN   

   Loans and advances  Government Securities  Total Deposits  \
0                 NaN                    NaN             NaN   
1                 NaN                    NaN             NaN   
2           6454927.0               850414.0       8394343.0   
3                 NaN                    NaN             NaN   
4                 NaN                    NaN             NaN   

   Required Liquidity  Actual Liquidity to Domestic Liabilities (%)  \
0                 NaN                                           NaN   
1                 NaN                                           NaN   
2            933050.0                                         0.219   
3                 NaN                                           NaN   
4                 NaN                                           NaN   

   Loans and Advances to Deposits(%)  Narrow Money Supply (M1)  \
0                                NaN                       NaN   
1                                NaN                       NaN   
2                              0.769                 2249220.0   
3                                NaN                       NaN   
4                                NaN                       NaN   

   Broad Money Supply (M2)  Currency outside depository corporations  \
0                      NaN                                       NaN   
1                      NaN                                       NaN   
2                8391420.0                                  356799.0   
3                      NaN                                       NaN   
4                      NaN                                       NaN   

   Demand deposits  Quasi Money  Inflation series        CPI  Discount Rate  \
0              NaN          NaN              4.35  58.987303            5.5   
1              NaN          NaN              4.19  59.633085            5.5   
2        1892421.0    6142199.0              5.53  60.125582            5.5   
3              NaN          NaN              6.65  60.699169            5.5   
4              NaN          NaN              7.11  61.152942            5.5   

   Banker's Call   
0            3.02  
1            3.13  
2            3.22  
3            3.26  
4            3.32

#### Display the shape of the dataset

In [16]:
# We will display the shape of the dataset
dataset.shape

(140, 18)

The are 140 rows in the dataset and 18 columns.

#### Print the information of the dataset

In [5]:
# The pandas dataframe.info() function is used to get a  concise summary of the dataframe
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140 entries, 0 to 139
Data columns (total 18 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   Date                                          140 non-null    object 
 1   Total liquidity                               122 non-null    float64
 2   Cash and Deposits with CBS                    122 non-null    float64
 3   Loans and advances                            122 non-null    float64
 4   Government Securities                         122 non-null    float64
 5   Total Deposits                                122 non-null    float64
 6   Required Liquidity                            122 non-null    float64
 7   Actual Liquidity to Domestic Liabilities (%)  122 non-null    float64
 8   Loans and Advances to Deposits(%)             122 non-null    float64
 9   Narrow Money Supply (M1)                      122 non-null    flo

#### Check for statistics

In [6]:
# The describe() method returns description of the data in the DataFrame.
dataset.describe()

Total liquidity  Cash and Deposits with CBS  Loans and advances  \
count     1.220000e+02                1.220000e+02        1.220000e+02   
mean      4.436439e+06                2.062406e+06        1.210677e+07   
std       2.071656e+06                1.115707e+06        2.375930e+06   
min       1.625736e+06                4.344260e+05        6.454927e+06   
25%       2.558157e+06                1.025086e+06        1.043978e+07   
50%       4.231372e+06                1.839369e+06        1.246728e+07   
75%       6.147074e+06                3.066568e+06        1.391257e+07   
max       8.804549e+06                4.460321e+06        1.627840e+07   

       Government Securities  Total Deposits  Required Liquidity  \
count           1.220000e+02    1.220000e+02        1.220000e+02   
mean            2.578700e+06    1.552155e+07        3.093704e+06   
std             1.110073e+06    4.520329e+06        1.111018e+06   
min             8.504140e+05    8.133019e+06        9.330500e+05   
25%             1.643543e+06    1.113150e+07        1.948417e+06   
50%             2.379200e+06    1.613529e+07        3.460694e+06   
75%             3.029590e+06    1.845494e+07        3.953601e+06   
max             4.787189e+06    2.390625e+07        7.229727e+06   

       Actual Liquidity to Domestic Liabilities (%)  \
count                                    122.000000   
mean                                       0.300631   
std                                        0.056761   
min                                        0.212000   
25%                                        0.255250   
50%                                        0.286500   
75%                                        0.350750   
max                                        0.431000   

       Loans and Advances to Deposits(%)  Narrow Money Supply (M1)  \
count                         122.000000              1.220000e+02   
mean                            0.803115              5.392919e+06   
std                             0.090350              1.585371e+06   
min                             0.617000              4.675340e+05   
25%                             0.742000              4.134178e+06   
50%                             0.798500              5.523949e+06   
75%                             0.876750              6.400437e+06   
max                             0.981000              8.548071e+06   

       Broad Money Supply (M2)  Currency outside depository corporations  \
count             1.220000e+02                                122.000000   
mean              1.534866e+07                             633962.803279   
std               3.924162e+06                             147264.574311   
min               8.391420e+06                             355902.000000   
25%               1.138945e+07                             527941.750000   
50%               1.660768e+07                             632877.000000   
75%               1.830702e+07                             727535.250000   
max               2.188661e+07                             986869.000000   

       Demand deposits   Quasi Money  Inflation series         CPI  \
count     1.220000e+02  1.220000e+02        126.000000  126.000000   
mean      4.799942e+06  9.917537e+06          5.585212   82.631361   
std       1.391317e+06  2.486023e+06          1.904830   13.286819   
min       1.892421e+06  5.878929e+06          1.600000   58.987303   
25%       3.621223e+06  7.337325e+06          4.525000   71.194676   
50%       4.902896e+06  1.082124e+07          5.327429   83.515767   
75%       5.659632e+06  1.192905e+07          6.695000   95.260121   
max       7.779620e+06  1.401826e+07          9.580000  103.870000   

       Discount Rate  Banker's Call   
count     128.000000      128.000000  
mean        5.699219        4.584539  
std         1.069145        1.490149  
min         3.750000        2.930000  
25%         5.000000        3.357500  
50%         5.500000        3.825000  
75%        

#### Evaluating Missing Values

In [7]:
# We will use .isnull().sum() to dertemine how many missing values exist in the collection, in which case you can use .sum() chained on.
dataset.isnull().sum()

Date                                             0
Total liquidity                                 18
Cash and Deposits with CBS                      18
Loans and advances                              18
Government Securities                           18
Total Deposits                                  18
Required Liquidity                              18
Actual Liquidity to Domestic Liabilities (%)    18
Loans and Advances to Deposits(%)               18
Narrow Money Supply (M1)                        18
Broad Money Supply (M2)                         18
Currency outside depository corporations        18
Demand deposits                                 18
Quasi Money                                     18
Inflation series                                14
CPI                                             14
Discount Rate                                   12
Banker's Call                                   12
dtype: int64

In [8]:
# Check the mean of all columns
dataset.mean()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  


Total liquidity                                 4.436439e+06
Cash and Deposits with CBS                      2.062406e+06
Loans and advances                              1.210677e+07
Government Securities                           2.578700e+06
Total Deposits                                  1.552155e+07
Required Liquidity                              3.093704e+06
Actual Liquidity to Domestic Liabilities (%)    3.006311e-01
Loans and Advances to Deposits(%)               8.031148e-01
Narrow Money Supply (M1)                        5.392919e+06
Broad Money Supply (M2)                         1.534866e+07
Currency outside depository corporations        6.339628e+05
Demand deposits                                 4.799942e+06
Quasi Money                                     9.917537e+06
Inflation series                                5.585212e+00
CPI                                             8.263136e+01
Discount Rate                                   5.699219e+00
Banker's Call           

## Fill Null Values

We will compare two methods of filling null values, namely the mean imputation and interpolate method.


##   Mean Imputation

In [9]:
# We use the mean to fill null values
data=dataset.fillna(dataset.mean())
data

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


Date  Total liquidity  Cash and Deposits with CBS  \
0     January 2011     4.436439e+06                2.062406e+06   
1    February 2011     4.436439e+06                2.062406e+06   
2       March 2011     1.625736e+06                4.344260e+05   
3       April 2011     4.436439e+06                2.062406e+06   
4         May 2011     4.436439e+06                2.062406e+06   
..             ...              ...                         ...   
135     April 2022     7.678791e+06                3.369518e+06   
136       May 2022     7.751584e+06                3.499925e+06   
137      June 2022     7.639919e+06                3.374612e+06   
138      July 2022     6.437907e+06                2.499161e+06   
139    August 2022     7.011310e+06                2.958616e+06   

     Loans and advances  Government Securities  Total Deposits  \
0          1.210677e+07           2.578700e+06    1.552155e+07   
1          1.210677e+07           2.578700e+06    1.552155e+07   
2          6.454927e+06           8.504140e+05    8.394343e+06   
3          1.210677e+07           2.578700e+06    1.552155e+07   
4          1.210677e+07           2.578700e+06    1.552155e+07   
..                  ...                    ...             ...   
135        1.565333e+07           4.708739e+06    2.141501e+07   
136        1.596195e+07           4.650874e+06    2.202960e+07   
137        1.627287e+07           4.613327e+06    2.254558e+07   
138        1.619444e+07           4.214094e+06    2.183174e+07   
139        1.627840e+07           4.350414e+06    2.200972e+07   

     Required Liquidity  Actual Liquidity to Domestic Liabilities (%)  \
0          3.093704e+06                                      0.300631   
1          3.093704e+06                                      0.300631   
2          9.330500e+05                                      0.219000   
3          3.093704e+06                                      0.300631   
4          3.093704e+06                                      0.300631   
..                  ...                                           ...   
135        3.905355e+06                                      0.389000   
136        3.957416e+06                                      0.387000   
137        4.094739e+06                                      0.369000   
138        3.932691e+06                                      0.323000   
139        3.980795e+06                                      0.348000   

     Loans and Advances to Deposits(%)  Narrow Money Supply (M1)  \
0                             0.803115              5.392919e+06   
1                             0.803115              5.392919e+06   
2                             0.769000              2.249220e+06   
3                             0.803115              5.392919e+06   
4                             0.803115              5.392919e+06   
..                                 ...                       ...   
135                           0.731000              7.534322e+06   
136                           0.725000              7.481177e+06   
137                           0.722000              7.894902e+06   
138                           0.742000              8.171439e+06   
139                           0.740000              7.778874e+06   

     Broad Money Supply (M2)  Currency outside depository corporations  \
0               1.534866e+07                             633962.803279   
1               1.534866e+07                             633962.803279   
2               8.391420e+06                             356799.000000   
3               1.534866e+07                             633962.803279   
4               1.534866e+07                             633962.803279   
..                       ...                                       ...   
135             1.988936e+07                             742614.000000   
136             2.036360e+07                             672836.000000   
137             2.088873e+07                            

#### 1. Dowload The Newly Created Dataset

In [ ]:
# We dowload the new dataset that was created after applying mean
data.to_excel('fixed_null_values.xlsx',index=False)

#### 2. Check Descriptive Statistics or Summary Statistics

In [15]:
# We get the description of the newly created dataset
data.describe()

Total liquidity  Cash and Deposits with CBS  Loans and advances  \
count     1.400000e+02                1.400000e+02        1.400000e+02   
mean      4.436439e+06                2.062406e+06        1.210677e+07   
std       1.932871e+06                1.040963e+06        2.216762e+06   
min       1.625736e+06                4.344260e+05        6.454927e+06   
25%       2.636046e+06                1.054516e+06        1.061183e+07   
50%       4.436439e+06                2.047408e+06        1.210677e+07   
75%       5.580148e+06                2.795640e+06        1.378658e+07   
max       8.804549e+06                4.460321e+06        1.627840e+07   

       Government Securities  Total Deposits  Required Liquidity  \
count           1.400000e+02    1.400000e+02        1.400000e+02   
mean            2.578700e+06    1.552155e+07        3.093704e+06   
std             1.035707e+06    4.217502e+06        1.036589e+06   
min             8.504140e+05    8.133019e+06        9.330500e+05   
25%             1.692328e+06    1.143652e+07        1.997177e+06   
50%             2.578700e+06    1.552155e+07        3.176190e+06   
75%             2.903214e+06    1.813742e+07        3.927101e+06   
max             4.787189e+06    2.390625e+07        7.229727e+06   

       Actual Liquidity to Domestic Liabilities (%)  \
count                                    140.000000   
mean                                       0.300631   
std                                        0.052958   
min                                        0.212000   
25%                                        0.257750   
50%                                        0.293000   
75%                                        0.332750   
max                                        0.431000   

       Loans and Advances to Deposits(%)  Narrow Money Supply (M1)  \
count                         140.000000              1.400000e+02   
mean                            0.803115              5.392919e+06   
std                             0.084297              1.479164e+06   
min                             0.617000              4.675340e+05   
25%                             0.750250              4.239612e+06   
50%                             0.803115              5.392919e+06   
75%                             0.853750              6.345806e+06   
max                             0.981000              8.548071e+06   

       Broad Money Supply (M2)  Currency outside depository corporations  \
count             1.400000e+02                                140.000000   
mean              1.534866e+07                             633962.803279   
std               3.661273e+06                             137398.999831   
min               8.391420e+06                             355902.000000   
25%               1.167230e+07                             540388.750000   
50%               1.534866e+07                             633962.803279   
75%               1.794863e+07                             704838.000000   
max               2.188661e+07                             986869.000000   

       Demand deposits   Quasi Money  Inflation series         CPI  \
count     1.400000e+02  1.400000e+02        140.000000  140.000000   
mean      4.799942e+06  9.917537e+06          5.585212   82.631361   
std       1.298109e+06  2.319479e+06          1.806358   12.599945   
min       1.892421e+06  5.878929e+06          1.600000   58.987303   
25%       3.710276e+06  7.422388e+06          4.675000   71.510470   
50%       4.799942e+06  9.917537e+06          5.582606   82.631361   
75%       5.574100e+06  1.179082e+07          6.562500   94.785303   
max       7.779620e+06  1.401826e+07          9.580000  103.870000   

       Discount Rate  Banker's Call   
count     140.000000      140.000000  
mean        5.699219        4.584539  
std         1.021953        1.424374  
min         3.750000        2.930000  
25%         5.000000        3.467500  
50%         5.500000        3.885000  
75%        

## Interpolation Method

In [11]:
# We first use the linear method for the interpolate function. The direction for interpolation is the forward direction
forward_data=dataset.fillna(dataset.interpolate(method='linear',limit_direction='forward'))

# Since there were still null values in the first rows of the dataset after applying the forward direction on the interpolate function, 
#  we then apply the backward direction on the newly cleaned data
backward_data=forward_data.fillna(forward_data.interpolate(method='linear',limit_direction='backward'))

#### 1. Print the new dataset that was created after using interpolate function

In [12]:
# Download the new dataset
new_data=backward_data.to_excel("interpolated_data.xlsx")

#### 2. Get the descriptive statistics of the data after applying the interpolate function

In [13]:
# Get the statistics from the newly created dataset after applying the interpolate function
backward_data.describe()

Total liquidity  Cash and Deposits with CBS  Loans and advances  \
count     1.400000e+02                1.400000e+02        1.400000e+02   
mean      4.123915e+06                1.896717e+06        1.158936e+07   
std       2.108410e+06                1.132405e+06        2.643805e+06   
min       1.625736e+06                4.344260e+05        6.454927e+06   
25%       2.352024e+06                9.667552e+05        9.393378e+06   
50%       3.598710e+06                1.508982e+06        1.189641e+07   
75%       5.580148e+06                2.795640e+06        1.378658e+07   
max       8.804549e+06                4.460321e+06        1.627840e+07   

       Government Securities  Total Deposits  Required Liquidity  \
count           1.400000e+02    1.400000e+02        1.400000e+02   
mean            2.404927e+06    1.470393e+07        2.890910e+06   
std             1.139847e+06    4.776327e+06        1.180879e+06   
min             8.504140e+05    8.133019e+06        9.330500e+05   
25%             1.504275e+06    1.040920e+07        1.813442e+06   
50%             2.208710e+06    1.424370e+07        3.188776e+06   
75%             2.903214e+06    1.813742e+07        3.927101e+06   
max             4.787189e+06    2.390625e+07        7.229727e+06   

       Actual Liquidity to Domestic Liabilities (%)  \
count                                    140.000000   
mean                                       0.293346   
std                                        0.056796   
min                                        0.212000   
25%                                        0.245000   
50%                                        0.279500   
75%                                        0.332750   
max                                        0.431000   

       Loans and Advances to Deposits(%)  Narrow Money Supply (M1)  \
count                         140.000000              1.400000e+02   
mean                            0.814204              5.077881e+06   
std                             0.093084              1.720132e+06   
min                             0.617000              4.675340e+05   
25%                             0.750250              3.839298e+06   
50%                             0.807000              4.931319e+06   
75%                             0.891000              6.345806e+06   
max                             0.981000              8.548071e+06   

       Broad Money Supply (M2)  Currency outside depository corporations  \
count             1.400000e+02                                140.000000   
mean              1.458556e+07                             603520.078571   
std               4.228013e+06                             160710.257420   
min               8.391420e+06                             355902.000000   
25%               1.083334e+07                             498197.750000   
50%               1.472583e+07                             594189.500000   
75%               1.794863e+07                             704838.000000   
max               2.188661e+07                             986869.000000   

       Demand deposits   Quasi Money  Inflation series         CPI  \
count     1.400000e+02  1.400000e+02        140.000000  140.000000   
mean      4.510077e+06  9.474382e+06          5.416691   84.755225   
std       1.529034e+06  2.625549e+06          1.876263   14.129680   
min       1.892421e+06  5.878929e+06          1.600000   58.987303   
25%       3.400899e+06  6.851690e+06          4.000000   71.510470   
50%       4.315552e+06  9.801496e+06          5.100000   86.138924   
75%       5.574100e+06  1.179082e+07          6.562500   96.747628   
max       7.779620e+06  1.401826e+07          9.580000  103.870000   

       Discount Rate  Banker's Call   
count     140.000000      140.000000  
mean        5.532143        4.504436  
std         1.159432        1.448371  
min         3.750000        2.930000  
25%         5.000000        3.467500  
50%         5.500000        3.660000  
75%        

To fill null values, we will choose the interpolate function with linear method. Interpolation is a technique in Python with which you can estimate unknown data points between two known data points. It is commonly used to fill missing values in a table or a dataset using the already known values.

The problem with mean imputation is that the mean imputation does not preserve the relationships among variables. 

Another problem with mean imputation is that it leads to an understimate of standard errors.

# 2. Exploratory Data Analysis